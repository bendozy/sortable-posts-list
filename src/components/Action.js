import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import timeTravel from '../actions/actions';

export const Action = ({
  action: { title, from, to },
  action,
  timeTravel: timeTravelAction,
  index,
}) => {
  // const [style, setStyle] = useState({
  //   opacity: 0,
  //   transition: 'all 2s ease',
  // });

  // useEffect(() => {
  //   setStyle({
  //     opacity: 1,
  //     transition: 'all 2s ease',
  //   });

  //   return () => {
  //     // setStyle({
  //     //   opacity: 1,
  //     //   transition: 'all 2s ease',
  //     // });
  //     setTimeout(() => {
  //       console.log('removing');
  //     }, 4000);
  //   };
  // }, [action]);

  return (
    <div
      className="Action bg-white rounded p-3 shadow border flex flex-wrap"
      // style={style}
    >
      <div className="sm:my-3 my-1 sm:text-base text-sm w-3/4 text-gray-700">
        {`Moved ${title} from index ${from} to index ${to}`}
      </div>
      <button
        onClick={() => timeTravelAction(index)}
        type="button"
        className="h-12 m-1 bg-green-200 rounded text-sm shadow sm:text-base w-1/5 text-gray-700"
      >
        Time Travel
      </button>
    </div>
  );
};

Action.propTypes = {
  action: PropTypes.shape({
    title: PropTypes.string.isRequired,
    from: PropTypes.number.isRequired,
    to: PropTypes.number.isRequired,
  }).isRequired,
  index: PropTypes.number.isRequired,
  timeTravel: PropTypes.func.isRequired,
};

const mapDispatchToProps = {
  timeTravel,
};

export default connect(null, mapDispatchToProps)(Action);
