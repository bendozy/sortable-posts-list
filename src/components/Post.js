import React from 'react';
import PropTypes from 'prop-types';
import { UP, DOWN } from '../constants';

// React Flip need the child component to not be stateless
/* eslint-disable react/prefer-stateless-function */
export default class Post extends React.Component {
  render() {
    const {
      post: { title },
      index,
      post,
      lastIndex,
      movePost: movePostAction,
    } = this.props;

    return (
      <div className="Post bg-white shadow rounded mb-6 flex justify-between">
        <div className="Post-title py-6 px-2 text-md">{title}</div>
        <div className="Post-buttons flex flex-col pr-2 justify-center">
          {index !== 0 && (
            <button
              type="button"
              className="Post-button moveUp"
              onClick={() => movePostAction({ direction: UP, post, index })}
            >
              <i className="text-2xl fa fa-angle-up" />
            </button>
          )}
          {index !== lastIndex && (
            <button
              type="button"
              className="Post-button moveDown"
              onClick={() => movePostAction({ direction: DOWN, post, index })}
            >
              <i className="text-2xl fa fa-angle-down" />
            </button>
          )}
        </div>
      </div>
    );
  }
}

Post.propTypes = {
  post: PropTypes.shape({
    title: PropTypes.string.isRequired,
  }).isRequired,
  index: PropTypes.number.isRequired,
  lastIndex: PropTypes.number.isRequired,
  movePost: PropTypes.func.isRequired,
};
