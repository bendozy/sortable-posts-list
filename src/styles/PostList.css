.PostList-spinner {
  min-height: 150px;
  background: #fff;
  margin: 0 auto;
  position: relative;
}

.PostList-cards {
  transform-style: preserve-3d;
}
