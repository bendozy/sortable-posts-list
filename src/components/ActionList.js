import React, { useRef, useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import {
  CSSTransition,
  TransitionGroup,
} from 'react-transition-group';
import Action from './Action';

export const ActionList = ({ actions }) => {
  const prevActionsRef = useRef([]);
  const [showLoading, setShowLoading] = useState(true);

  useEffect(() => {
    if (actions.length === 0 && prevActionsRef.current.length > actions.length) {
      setTimeout(() => {
        setShowLoading(true);
      }, 500);
    } else if (prevActionsRef.current.length === 0 && actions.length > 0) {
      setShowLoading(false);
    }

    prevActionsRef.current = actions;
  }, [actions]);

  return (
    <div className="ActionList">
      <div className="bg-white rounded shadow-md overflow-hidden">
        <div className="ActionList-title font-semibold bg-white p-4 text-gray-800 text-xl">List of actions commited</div>
        {actions.length > 0 && (
          <div className="ActionList-items bg-gray-300 p-6 border border-gray-300 overflow-hidden shadow-md">
            <TransitionGroup className="action-list">
              {actions.map((action, index) => (
                <CSSTransition
                  key={action.key}
                  timeout={500}
                  classNames="item"
                >
                  <Action
                    action={action}
                    index={index}
                  />
                </CSSTransition>
              )).reverse()}
            </TransitionGroup>
          </div>
        )}
        {actions.length === 0 && showLoading && (
          <div className="ActionList-loading bg-white p-4 rounded shadow-md text-gray-500">
            <span>No actions commited</span>
          </div>
        )}
      </div>
    </div>
  );
};

ActionList.propTypes = {
  actions: PropTypes.arrayOf(PropTypes.object).isRequired,
};

export const mapStateToProps = ({ actions: { actions } }) => ({ actions });

export default connect(mapStateToProps)(ActionList);
