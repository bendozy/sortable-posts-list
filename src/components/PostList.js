import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import {
  CSSTransition,
  TransitionGroup,
} from 'react-transition-group';
import { connect } from 'react-redux';
import Post from './Post';
import { fetchPosts } from '../actions/posts';

import spinner from '../assets/Spinner.svg';

import '../styles/PostList.css';

export const PostList = ({
  posts: {
    isLoading,
    posts,
  },
  fetchPosts: fetchPostsAction,
}) => {
  useEffect(() => {
    fetchPostsAction();
  }, [fetchPostsAction]);


  const postCards = posts.map((post, index) => (
    <CSSTransition
      key={post.id}
      timeout={2000}
      classNames="item"
    >
      <Post
        post={post}
        index={index}
        lastIndex={posts.length - 1}
      />
    </CSSTransition>
  ));

  return (
    <div className="PostList">
      <h1 className="mb-6 font-bold text-3xl text-white">Sortable Post List</h1>
      {isLoading && <img src={spinner} className="PostList-spinner" alt="loading-logo" />}
      {!isLoading && (
        <div className="PostList-cards">
          <TransitionGroup className="todo-lists">
            {postCards}
          </TransitionGroup>
        </div>
      )}
    </div>
  );
};

PostList.propTypes = {
  fetchPosts: PropTypes.func.isRequired,
  posts: PropTypes.shape({
    isLoading: PropTypes.bool.isRequired,
    posts: PropTypes.arrayOf(PropTypes.object),
  }).isRequired,
};

export const mapStateToProps = ({ posts }) => ({ posts });

const mapDispatchToProps = {
  fetchPosts,
};

export default connect(mapStateToProps, mapDispatchToProps)(PostList);
