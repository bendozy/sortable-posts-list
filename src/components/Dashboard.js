import React from 'react';
import PostList from './PostList';
import ActionList from './ActionList';

const Dashboard = () => (
  <div id="Dashboard">
    <PostList />
    <ActionList />
  </div>
);

export default Dashboard;
