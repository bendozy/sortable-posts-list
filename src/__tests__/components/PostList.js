import React from 'react';
import { act } from 'react-dom/test-utils';
import toJson from 'enzyme-to-json';
import { PostList, mapStateToProps } from '../../components/PostList';
import Post from '../../components/Post';

const fetchPosts = jest.fn();

jest.mock('../../components/Post', () => () => 'Post');
jest.mock('react-redux', () => ({
  connect: () => (wrappedComponent) => wrappedComponent,
}));

describe('<PostList />', () => {
  it('should render a <PostList /> component', () => {
    const posts = {
      posts: [],
      isLoading: true,
    };
    const wrapper = shallow(<PostList fetchPosts={fetchPosts} posts={posts} />);

    expect(wrapper.find(Post)).toHaveLength(0);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should update a <PostList /> component', () => {
    const posts = {
      posts: [],
      isLoading: true,
    };
    const wrapper = mount(<PostList fetchPosts={fetchPosts} posts={posts} />);

    expect(wrapper.find(Post)).toHaveLength(0);

    act(() => {
      wrapper.setProps({
        posts: {
          posts: [
            { id: 'test1' },
            { id: 'test2' },
          ],
          isLoading: false,
        },
      });
    });
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should return correct props from mapStateToProps', () => {
    const posts = [
      { id: 'test1' },
      { id: 'test2' },
    ];

    const initialState = {
      posts: {
        posts,
        isLoading: true,
      },
    };

    const result = mapStateToProps(initialState);

    expect(result).toEqual(initialState);
    expect(toJson(result)).toMatchSnapshot();
  });
});
