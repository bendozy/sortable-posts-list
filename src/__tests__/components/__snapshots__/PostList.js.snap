// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PostList /> should render a <PostList /> component 1`] = `
<div
  className="PostList"
>
  <h1
    className="PostList-titlemb-6 font-bold text-3xl text-white"
  >
    Sortable Post List
  </h1>
  <img
    alt="loading-logo"
    className="PostList-spinner"
    src="Spinner.svg"
  />
</div>
`;

exports[`<PostList /> should return correct props from mapStateToProps 1`] = `undefined`;

exports[`<PostList /> should update a <PostList /> component 1`] = `
<PostList
  fetchPosts={
    [MockFunction] {
      "calls": Array [
        Array [],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  posts={
    Object {
      "isLoading": false,
      "posts": Array [
        Object {
          "id": "test1",
        },
        Object {
          "id": "test2",
        },
      ],
    }
  }
>
  <div
    className="PostList"
  >
    <h1
      className="PostList-titlemb-6 font-bold text-3xl text-white"
    >
      Sortable Post List
    </h1>
    <div
      className="PostList-items"
    >
      <TransitionGroup
        childFactory={[Function]}
        className="todo-lists"
        component="div"
      >
        <div
          className="todo-lists"
        >
          <CSSTransition
            classNames="item"
            in={true}
            key=".$test1"
            onExited={[Function]}
            timeout={2000}
          >
            <Transition
              appear={false}
              enter={true}
              exit={true}
              in={true}
              mountOnEnter={false}
              onEnter={[Function]}
              onEntered={[Function]}
              onEntering={[Function]}
              onExit={[Function]}
              onExited={[Function]}
              onExiting={[Function]}
              timeout={2000}
              unmountOnExit={false}
            >
              <Component
                index={0}
                lastIndex={1}
                post={
                  Object {
                    "id": "test1",
                  }
                }
              >
                Post
              </Component>
            </Transition>
          </CSSTransition>
          <CSSTransition
            classNames="item"
            in={true}
            key=".$test2"
            onExited={[Function]}
            timeout={2000}
          >
            <Transition
              appear={false}
              enter={true}
              exit={true}
              in={true}
              mountOnEnter={false}
              onEnter={[Function]}
              onEntered={[Function]}
              onEntering={[Function]}
              onExit={[Function]}
              onExited={[Function]}
              onExiting={[Function]}
              timeout={2000}
              unmountOnExit={false}
            >
              <Component
                index={1}
                lastIndex={1}
                post={
                  Object {
                    "id": "test2",
                  }
                }
              >
                Post
              </Component>
            </Transition>
          </CSSTransition>
        </div>
      </TransitionGroup>
    </div>
  </div>
</PostList>
`;
